nntrainer/layers/pooling2d_layer.cpp

```diff
@@ -244,57 +277,104 @@ void Pooling2DLayer::setProperty(const PropertyType type,
   }
 }
 
-Tensor Pooling2DLayer::pooling2d(unsigned int batch, Tensor &in,
-                                 Tensor &output) {
+Tensor Pooling2DLayer::pooling2d(Tensor &in, bool training, Tensor &output) {
+
   unsigned int channel = in.channel();
-  unsigned int height = in.height();
-  unsigned int width = in.width();
-  unsigned int p_height = pool_size[0];
-  unsigned int p_width = pool_size[1];
-  TensorDim &out_dim = output_dim[0];
-  unsigned int base_idx = batch * out_dim.getFeatureLen();
+  unsigned int pad_height = padding[0];
+  unsigned int pad_width = padding[1];
+  int in_height = in.height();
+  int in_width = in.width();
+  unsigned int height = in_height + pad_height * 2;
+  unsigned int width = in_width + pad_width * 2;
+  unsigned int patch_height = pool_size[0];
+  unsigned int patch_width = pool_size[1];
+
+  NNTR_THROW_IF(output.uninitialized(), std::invalid_argument)
+    << "[Pooling2D] output is uninitialized, this is not supported";
+
+  /**
+   * @brief pooling function
+   * @param in_c channel sliced data
+   * @param start_h (height index pointing the start of the patch)
+   * @param start_w (width index pointing the start of the patch)
+   * @return result value of pooling
+   */
+  std::function<float(const float *, int, int)> pool_fn;
 
-  if (output.uninitialized())
-    output = Tensor(1, out_dim.channel(), out_dim.height(), out_dim.width());
+  switch (pooling_type) {
+  case PoolingType::max: {
+    pool_fn = [&, this](const float *in_data, int start_h, int start_w) {
+      int end_h = start_h + patch_height;
+      int end_w = start_w + patch_width;
+
+      float max_val = std::numeric_limits<float>::lowest();
+
+      int cur_max_idx = -1;
+      for (int h = start_h; h < end_h; ++h) {
+        if (h < 0 || in_height <= h) {
+          continue;
+        }
+        for (int w = start_w; w < end_w; ++w) {
```

Comment:
`for (int w = std::max(0, start_w); w < std::min(end_w, in_width); ++w)` and remove the check below?

Reply:
Done